Added `Notification-Service` for contract events (#240)
sidestream-tech/unified-auctions-ui: notification-service/README.md

- `ETHEREUM_NETWORK`: (optional, default `kovan`) – internal network name on which the bot poll for auctions. Available
- `SMTP_*`: (optional, required for email usage) - SMTP data for the service to send emails
    - `SMTP_HOST`: (required) - SMTP host address
    - `SMTP_PORT`: (optional) - SMTP port. Defaults to `465`
    - `SMTP_USERNAME`: (required) - SMTP username
    - `SMTP_PASSWORD`: (required) - SMTP password
    - `SMTP_EMAIL`: (required) - The outgoing address which is displayed to the user
- `RECEIVERS`: (required) - a json-formatted object with `receiver` as key and `Event Subscription Ids` as an array value, eg `RECEIVERS="{\"email@example.com\": [\"EVENT_1\", \"EVENT_2\"]}"`. The email will then become subscribed to the listed events

Note: env variables are accessible via the `secret` command under `auction-ui/notification-service/${environment}`.

## Adding a new "Receiver"
You can set your receivers over the environment variable `RECEIVERS`. This is a json-formatted object. An example can be found below:

```json
{
  "test@example.com": {
    "type": "email",
    "subscriptions": ["MCD_DAI_Transfer", "ChainLogUpdateAddress"]

>   },

Reviewer: Suggested format was an _array_ of objects. Why? Because eventually this data will be stored in a database instead of the env variable

Author: I had actually wanted to do this but for some reason it was ingrained into my brain that the issue stated the other. Happy to change it tho!

- `ETHEREUM_NETWORK`: (optional, default `kovan`) – internal network name on which the bot poll for auctions. Available
- `SMTP_*`: (required) - SMTP data for the service to send emails
    - `SMTP_HOST`: (required) - SMTP host address
    - `SMTP_PORT`: (optional) - SMTP port. Defaults to `465`
    - `SMTP_USERNAME`: (required) - SMTP username
    - `SMTP_PASSWORD`: (required) - SMTP password
    - `SMTP_EMAIL`: (optional) - The displayed address from which the email was sent. Defaults to `notification@unified-auctions.makerdao.com`
- `RECEIVERS`: (optional) - a comma-separated list of email addresses, that should receive an email when an event is
  called

Note: env variables are accessible via the `secret` command under `auction-ui/notification-service/${environment}`.

## Adding a new "Event Subscription"

In order to add a custom event subscription you need to edit the file `/constants/SUBSCRIPTIONS.ts`. Inside you will
find an array of Event Subscriptions. You can add your own with the following format:

```js
export const SUBSCRIPTIONS: EventSubscription[] = [
  {
    id: 'ChainlogUpdateVersion',
    contract: '0xdA0Ab1e0017DEbCd72Be8599041a2aa3bA7e740F',

>     abi: CHAINLOG_ABI,

Reviewer: Can we please fetch this from etherscan? https://docs.etherscan.io/api-endpoints/contracts#get-contract-abi-for-verified-contract-source-codes

Author: Do we have an API for this? 

https://docs.etherscan.io/getting-started/viewing-api-usage-statistics
I can create my own for free, however we should look into generating one for production then.

Author: In addition to this. Don't we want to implement this into `/core` as well then? Or is there some kind of caveat (aside from the rate limiting and generally us having to use more api calls)?

Reviewer: > Don't we want to implement this into `/core` as well then?

No, as specified in the issue discussion, we don't share code in this service with the `core`

$ npm install

# serve with hot reload
$ npm run dev

# build for production and launch server
$ npm run build
$ npm run start
```

## Environment variables

- `INFURA_PROJECT_ID`: (required) [infura](https://infura.io/) project id (can be found in: dashboard -> ethereum ->
- `ETHERSCAN_API_KEY`: (required) [Etherscan API Key](https://docs.etherscan.io/getting-started/viewing-api-usage-statistics), which is used to automatically download ABIs 
- `ETHEREUM_NETWORK`: (optional, default `kovan`) – internal network name on which the bot poll for auctions. Available
- `SMTP_*`: (required) - SMTP data for the service to send emails
    - `SMTP_HOST`: (required) - SMTP host address
    - `SMTP_PORT`: (optional) - SMTP port. Defaults to `465`
    - `SMTP_USERNAME`: (required) - SMTP username
    - `SMTP_PASSWORD`: (required) - SMTP password
    - `SMTP_EMAIL`: (required) - The outgoing address which is displayed to the user
- `RECEIVERS`: (required) - a json-formatted object with `email` as key and `Event Subscription Ids` as an array value, eg `RECEIVERS="{\"email@example.com\": [\"EVENT_1\", \"EVENT_2\"]}"`. The email will then become subscribed to the listed events

> Note: env variables are accessible via the `secret` command under `auction-ui/notification-service/${environment}`.

Reviewer: The path to the secrets should be `auction-ui/${host}/notification-service`. 

But I couldn't find any secrets neither under `auction-ui/notification-service/localhost` or under `auction-ui/notification-service/dev`. Where are they currently stored?

Author: I have not added any secrets yet, as we need to generate new API tokens for etherscan. Who should take care of this?

Reviewer 2: I took that one over and created an API key for us. You can find credentials in bitwarden under "SAS Etherscan Account"

Author: Done! ✅ Also changed in the bot readme

## Adding a new "Event Subscription"

In order to add a custom event subscription you need to edit the file `/constants/SUBSCRIPTIONS.ts`. Inside you will
find an array of Event Subscriptions. You can add your own with the following format:

```js
export const SUBSCRIPTIONS: EventSubscription[] = [
  {
    id: 'ChainlogUpdateVersion',
    contract: '0xdA0Ab1e0017DEbCd72Be8599041a2aa3bA7e740F',
    eventName: 'UpdateVersion',
    formatData: event => {
      return `Key: ${event.args.key}<br /> Address: ${event.args.addr}`;
    },
  }
]
```

> - `id` - Unique identifier. Used for references in email and console logs

Reviewer: Can we use `contract:eventName` as unique identifier in logs? Or is it actually used for some other purposes?

Author: It is primarily needed at the id that emails can subscribe to. But I will also use it in the logs!

> - `contract` - The Ethereum Address on which your Event is

Reviewer: I would call it `address`
```suggestion
- `address` - The Ethereum address which expected to emit the event
```

> - `eventName` - The name of the event you want to observe

Reviewer: If you support wildcard, you should specify it here
```suggestion
- `eventName` - The name of the event you want to observe or `*`
```

Author: Using `*` does not work anymore, due to me changing the way the websockets are triggered. Therefore I removed this from the Readme.

context:
Before I listened to the event name, `*`, exists as an event name, but now I listen to all events be default and then compare their name (event.event) to the event listed in the Subscription. Therefore `*` would not work unless I support it again, which I don't want to do, as it could mess up with `formatData()` function, if events have different args.

> - `formatData` - A function that takes the event output, can reformat and then output it in html. It will be displayed in the email.

Reviewer: ```suggestion
- `formatData` - A function that takes the emitted event data and have to return an html string that will be displayed in the email.
```

## Development Setup

Please see [this centralized guide](https://github.com/sidestream-tech/guides/blob/main/frontend-development/README.md)
to get started with your development setup. Namely, you should:

- have a minimum `node` and `npm` version (see `package.json` `engine` section)
- have a certain formatting and linting setup
- don't forget to create `./notification-service/.env` file

Help on both things is given in the linked resources above.

In addition to this please be aware of the following for this project:

> - By default, this project uses [Etheral Email](https://ethereal.email/) to mock emails. If Etheral is down and no custom SMTP server is set, the service will not run.

Reviewer: What means by default? If no `SMTP_` variables are provided? Then why not add this note to the env variable description itself?

Author: `nodemailer`, the package I use to send emails, comes with a testing SMTP server built in. 

See more here:
https://nodemailer.com/smtp/testing/

I thought it would be easy to include this, to do a quick test without having to setup your own SMTP server or MailTrap. However if you think it is not needed, I can remove the feature and have the service fail, if no valid SMTP server is provided.

Author: Not relevant anymore as I removed this feature!

> - We advise using [MailTrap](https://mailtrap.io/) for more extensive testing. 

Reviewer: I don't think we need to advertise any specific email service in our readme
